### ISSUE TITLE
Kotlin lazy-initialization 

### ISSUE THREAD

**Issue by ythy** [ythy/blog#144]
`by lazy { ... }` performs its initializer where the defined property is first used, not its declaration.

```
class MainActivity : AppCompatActivity() {
    private val messageView : TextView by lazy {
        // runs on first access of messageView
        findViewById(R.id.message_view) as TextView
    }
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
    }
    fun onSayHello() {
        // Initialization would be run at here!!
        messageView.text = "Hello"
    }
}
```
Now, we can declare a read-only property without worrying about the initialization point of messageView.

### Delegated property 
Delegation literally means delegation. It means the delegator can perform some operation instead the original accessor.

Delegated property delegates a getter/setter of the property, which allows the delegated object to perform some intermediate action when reading and writing the value.

You can declare a delegated property in the following by `<delegate>` format :
```
val / var <property name>: <Type> by <delegate>
```

### How ‘by lazy’ works
 
If you do not specify a particular mode, lazy implementations would be `SynchronizedLazyImpl` which performs the initialization only once by default . Let's look at its implementation code .

```kotlin
private object UNINITIALIZED_VALUE
private class SynchronizedLazyImpl<out T>(
        initializer: () -> T,
        lock: Any? = null
) : Lazy<T>,
    Serializable {
    private var initializer: (() -> T)? = initializer
    @Volatile private var _value: Any? = UNINITIALIZED_VALUE
    // final field is required
    // to enable safe publication of constructed instance
    private val lock = lock ?: this
    override val value: T
        get() {
            val _v1 = _value
            if (_v1 !== UNINITIALIZED_VALUE) {
                @Suppress("UNCHECKED_CAST")
                return _v1 as T
            }
            return synchronized(lock) {
                val _v2 = _value
                if (_v2 !== UNINITIALIZED_VALUE) {
                    @Suppress("UNCHECKED_CAST") (_v2 as T)
                }
                else {
                    val typedValue = initializer!!()
                    _value = typedValue
                    initializer = null
                    typedValue
                }
            }
        }
    override fun isInitialized(): Boolean =
            _value !== UNINITIALIZED_VALUE
    override fun toString(): String =
            if (isInitialized()) value.toString()
            else "Lazy value not initialized yet."
    private fun writeReplace(): Any =
            InitializedLazyImpl(value)
}
```

It seems to be a bit complicated. but it’s just the same implementation for multi-threading.
* Executing the initialization block with `synchronized()`.
* Since the initialization may already be finished by another thread, it does **double-check**. If initialization already has been done, it returns the stored value.
* If it’s uninitialized, it will execute the lambda expression and store the return value. Theninitializer will be null, because it’s no longer needed upon initialization completion .

概述：
```
class Demo { 
    val myName: String by lazy {  getNameFromPreference() }
 }
```
1. 此处 ` getNameFromPreference() ` 为 `initializer` 
2. 初始   `_value` 为 `UNINITIALIZED_VALUE`
3. 当第一次读取 `myName`时， 如果`_value`为`UNINITIALIZED_VALUE`, 执行`initializer`, `_value`
设为`getNameFromPreference()`的返回值， 即`myName`的值。`initializer` 置为null
4. 再次读取`myName`时，`_value`不等于`UNINITIALIZED_VALUE`, 继续返回当前`_value`
